packages/yarnpkg-pnp/sources/esm-loader/loader.ts

```diff
@@ -0,0 +1,218 @@
+
+const cachedFS = new CachedInputFileSystem(fs);
+
+function isDirectory(filePath: string) {
+  return new Promise<boolean>(resolve => {
+    cachedFS.lstat!(filePath, (err, stat) => {
+      if (err || !stat) {
+        resolve(false);
+      } else {
+        resolve(stat.isDirectory());
+      }
+    });
+  });
+}
+
+function readFile(filePath: string) {
+  return new Promise<string>((resolve, reject) => {
+    cachedFS.readFile(filePath, (err, result) => {
+      if (err || !result) {
+        reject(err);
+      } else {
+        resolve(Buffer.isBuffer(result) ? result.toString(`utf8`) : result);
+      }
+    });
+  });
+}
+
+function readJson(filePath: string) {
+  return new Promise<any>((resolve, reject) => {
+    cachedFS.readJson!(filePath, (err, result) => {
+      if (err) {
+        reject(err);
+      } else {
+        resolve(result);
+      }
+    });
+  });
+}
+
+const commonResolver = ResolverFactory.createResolver({
+  fileSystem: cachedFS,
+  conditionNames: [`node`, `import`],
+  extensions: [`.js`, `.json`],
+  // TODO: Disable if --experimental-specifier-resolution=node
+  fullySpecified: true,
+});
+
+// @ts-expect-error - This module, when bundled, is still ESM so this is valid
+const pnpapi: PnpApi = createRequire(import.meta.url)(`pnpapi`);
+
+// TODO: Reuse this from the other files (this is the third copy)
+const pathRegExp = /^(?![a-zA-Z]:[\\/]|\\\\|\.{0,2}(?:\/|$))((?:@[^/]+\/)?[^/]+)\/*(.*|)$/;
+
+export async function resolve(specifier: string, context: any, defaultResolver: any) {
+  let validURL;
+  if (builtins.has(specifier) || (validURL = isValidURL(specifier))) {
+    if (!validURL || pathToFileURL(specifier).protocol !== `file:`) {
+      return defaultResolver(specifier, context, defaultResolver);
+    } else {
+      specifier = fileURLToPath(specifier);
+    }
+  }
+
+  const {parentURL, conditions = []} = context;
+
+  const resolver =
+    conditions.join(`.`) === `node.import`
+      ? commonResolver
+      : ResolverFactory.createResolver({
+        fileSystem: cachedFS,
+        conditionNames: conditions,
+        extensions: [`.js`, `.json`],
+        // TODO: Disable if --experimental-specifier-resolution=node
+        fullySpecified: true,
+      });
+
+  let parentPath = parentURL ? fileURLToPath(parentURL) : process.cwd();
+  try {
+    if (specifier.startsWith(`.`) && !(await isDirectory(parentPath))) {
+      parentPath = path.dirname(parentPath);
+    }
+  } catch {}
+
+  return new Promise((resolve, reject) => {
+    resolver.resolve({}, parentPath, specifier, {}, (err, file) => {
+      if (err || !file) {
+        reject(err);
+      } else {
+        // Guard against https://github.com/webpack/enhanced-resolve/issues/273
+        if (specifier.match(pathRegExp) !== null && !pnpapi.findPackageLocator(file)) {
+          reject(new Error(`'enhanced-resolve' went outside of the pnpapi and resolved '${specifier}' to '${file}'`));
+        } else {
```

Comment:
`enhanced-resolve` happily leaves the package path given to it by PnP so need to prevent that from happening

ref https://github.com/webpack/enhanced-resolve/issues/273
cc @sokra